README.md
Install
=======

pip install --user requests[security]

(try removing --user and install with sudo in case of errors)

git clone --depth=1 https://github.com/deibit/cansina


News
====

[CHANGELOG](https://github.com/deibit/cansina/blob/master/CHANGELOG.md)


Cansina
=======

[![Build Status](https://travis-ci.org/deibit/cansina.svg?branch=master)](https://travis-ci.org/deibit/cansina)

Cansina is a Web Content Discovery Application.

It is well known Web applications don't publish all their resources or public links, 
so the only way to discover these resources is requesting for them and check the response.

Cansina duty is to help you making requests and filtering and inspecting the responses to tell 
apart if it is an existing resource or just an annoying or disguised 404.

Feature requests and comments are welcome.

Cansina is included in [BlackArch Linux](https://www.blackarch.org/), give it a try! 

Screenshot
----------

![CansinaImage](https://github.com/deibit/cansina/raw/gh-pages/images/cansina-showcase.png "Image")

Features
--------

- Multithreading
- SSL support
- Proxy support
- Data persistence with sqlite database
- Basic Authentication
- Cookie jar
- Resuming
- Path recursion
- Persistent connections
- Complementary tools


Usage
-----

cansina.py -h for a comprehensive list of features and choices.

Look up the [wiki](https://github.com/deibit/cansina/wiki), it's full of information.


Important
---------

This tool is intended to be used in a fair and legal context, meaning, for example,
a penetration testing for which you have been provided previous authorization.

One of its legitimate uses might be the one described in the following article:

- [Forced browsing](https://www.owasp.org/index.php/Forced_browsing) 


Dependencies
------------

- [requests](https://github.com/kennethreitz/requests)
- Python 2.7.x or Python 3


Payloads
--------

Cansina does not come with list but there are some neat projects to supply this:

- [SecList](https://github.com/danielmiessler/SecLists)
- [fuzzdb](https://code.google.com/p/fuzzdb/)


License information
-------------------

Copyright (C) 2013-2017 David García

License: GNU General Public License, version 3 or later; see LICENSE.txt
         included in this archive for details.


CHANGELOG.md
Changelog
=========

15-09-2017  Payloads can be now a single file, a directory or a file containing references to other files

14-09-2017  New console interface and new option 'show-type' to reflect 'Content-Type' in output

11-09-2017  Added a new option 'full-path' for show the entire path in console

07-09-2017  Added an 'eta' clock 

06-09-2017  Viewer (from utils/viewer.py) received some love. A tree struct view and a
            couple of filters. Give it a try (utils/viewer.py -h)

03-09-2017  Added persistent connections (--persist). You will definitely want this.

31-08-2017  Added recursive (--recursive) feature

31-08-2017  Support for Python 3 has been tested and seems stable

22-08-2017  Cookie support. With option -C you can now provide a set of cookie like "JSESSIONID:blablabla,admin=1"

14-08-2017  Size filtering. Now can be a list of sizes you want to ignore

13-03-2017  New option: -R autoscan robots.txt and use it as a payload

26-01-2017  Fixed 'InsecureRequestsWarning' message

10-08-2016  Added '-u' option to utils/viewer.py to show what payload you used in a given project

28-07-2016  Fixed max columns width when printing output

19-07-2016  Fixed a bug in DBManager queue which was stopping Cansina when target server responses were slow

13-07-2016  Added new feature resumable sessions

09-01-2016  Cansina supports Windows terminal
            You won't have that fancy flash in filtered HTTP codes but it works...    


License information
-------------------

Copyright (C) 2013-2017 David García

License: GNU General Public License, version 3 or later; see LICENSE.txt
         included in this archive for details.


core/payload.py
import sys
import threading
import time
import fileinput
import os
import os.path
import fnmatch

try:
    import Queue
except:
    import queue as Queue

try:
    import urlparse
except:
    import urllib.parse as urlparse

from core.task import Task

def _get_url_components(target):
    ''' Get a url and returns multiple paths composed by recursive components'''
    path = [i for i in urlparse.urlparse(target).path.split('/') if len(i) > 0]
    temp_path = ""
    list_path = ['/']
    for component in path:
        sub = "/%s" % component
        temp_path += sub
        list_path.append(temp_path + '/')
    return list_path


def _populate_list_with_file(file_name, linenumber):
    """ Open a file, read its content and strips it. Returns a list with the content
        additionally it filter and clean some splinters
    """
    def _read_a_file_return_a_list(file_name):
        '''
            Small utility function. Open a file a return its content as a list
        '''
        tmp_list = []
        try:
            if sys.version_info > (3, 0):
                f =  open(file_name, 'r', encoding="latin-1")
            else:
                f =  open(file_name, 'r')
            tmp_list = f.readlines()
        except (OSError, IOError) as e:
            print("[!] Opening payload. Check file, list of files or directory content.")
            print(e)
            sys.exit()
        return tmp_list

    #FIXME: Garbage
    tmp_list = None
    # If we receive a list then they are robots.txt entries
    if type(file_name) == list:
        tmp_list = file_name
    # If filename is - we are dealing with stdin
    elif file_name == '-':
        tmp_list = []
        for line in fileinput.input(files=file_name):
            tmp_list.append(line)
    # Single file, could be a list of lists or a single payload
    elif os.path.isfile(file_name):
        tmp_list = _read_a_file_return_a_list(file_name)
        # Now is time to check if is a list or single payload
        if not file_name.endswith('.payload'):
            tmp_list = tmp_list[linenumber:]
        else:
            payload_list = tmp_list
            tmp_list = []
            print("[*] Trying to load {0} files as payloads".format(len(payload_list)))
            for item in payload_list:
                tmp_list.extend(_read_a_file_return_a_list(item.strip()))

    # If we receive a dir then try to open all .txt files in there
    elif os.path.isdir(file_name):
        tmp_list = []
        print("[*] Directory payload inclusion. All *.txt will be treated as a payload.")
        for file in os.listdir(file_name):
            if fnmatch.fnmatch(file, '*.txt'):
                tmp_list.extend(_read_a_file_return_a_list("{0}/{1}".format(file_name, file)))

    clean_list = []
    for e in tmp_list:
        # Delete leading and trailing spaces
        e = e.strip()
        # Skip commented lines in payload files
        if e.startswith('#'):
            continue
        # Remove leading '/' characters
        if e.startswith('/'):
            e = e[1:]

        if sys.version_info[0] == 3:
            e_encode = e
        else:
            e_encode = e.decode('utf-8', 'replace')
        clean_list.append(e_encode)
    return clean_list

def _has_extension(res):
    #whether the last path sector has '.'
    if res.rfind("/") == -1:
        return "." in res
    else:
        return "." in res[res.rfind("/"):]

class Payload():
    def __init__(self, target, payload_filename, resumer):

        self.target = target
        self.task_id = 0
        # Payload may be a filename, a python list, a directory or a file with
        # paths to multiple payload files
        self.payload_filename = payload_filename if not type(payload_filename) == list else "robots.txt"
        self.linenumber = resumer.get_line()
        self.payload = _populate_list_with_file(payload_filename, self.linenumber)
        self.queue = Queue.Queue()
        self.dead = False
        self.extensions = None
        self.banned_response_codes = None
        self.unbanned_response_codes = None
        self.content = None
        self.remove_slash = False
        self.uppercase = False
        self.recursive = False
        self.total_requests = 0

    def get_total_requests(self):
        return self.total_requests

    def set_recursive(self, recursion):
        self.recursive = recursion

    def set_remove_slash(self, remove_slash):
        self.remove_slash = remove_slash

    def set_banned_response_codes(self, banned_response_codes):
        self.banned_response_codes = banned_response_codes

    def set_unbanned_response_codes(self, unbanned_response_codes):
        self.unbanned_response_codes = unbanned_response_codes

    def set_extensions(self, extensions):
        self.extensions = extensions

    def set_content(self, content):
        self.content = content

    def kill(self):
        self.dead = True

    def is_finished(self):
        return self.dead

    def set_uppercase(self, uppercase):
        self.uppercase = uppercase

    def _feed_queue(self):
        for resource in self.payload:
            if self.uppercase:
                resource = resource.upper()

            # Useful when looking for files without extension instead of directories
            if self.remove_slash and resource.endswith("/"):
                resource = resource[:-1]

            for extension in self.extensions:
                # If resource is a whole word and user didnt provide a extension
                # put a final /
                if not extension and not _has_extension(resource) and not self.remove_slash:
                    resource += '/'

                # Put a . before extension if the users didnt do it
                if extension and not '.' in extension:
                    extension = '.' + extension

                self.task_id += 1
                task = Task(self.task_id, self.target, resource, extension)
                task.set_payload_filename(self.payload_filename)
                task.set_payload_length(self.get_total_requests())
                task.set_banned_response_codes(self.banned_response_codes)
                task.set_unbanned_response_codes(self.unbanned_response_codes)
                task.set_content(self.content)
                self.queue.put(task)

    def get_queue(self):
        if self.recursive:
            url_components = _get_url_components(self.target)
            # save main component (scheme + netloc)
            path = urlparse.urlparse(self.target).path
            saved_main_component = self.target[:self.target.find(path)]
            self.total_requests = len(self.payload) * len(self.extensions) * len(url_components)

            # if user select recursive but url is just root
            if len(url_components) == 1:
                self._feed_queue()
            else:
                # iterate over generated multiple paths
                for i in url_components:
                    self.target = saved_main_component + i
                    self._feed_queue()
        else:
            self.total_requests = len(self.payload) * len(self.extensions)
            self._feed_queue()
        return self.queue


core/printer.py
import os
import sys
import time

try:
    import urlparse
except:
    import urllib.parse as urlparse

if os.name == 'nt':
    RED,MAGENTA,BLUE,GREEN,YELLOW,LBLUE,ENDC = ("","","","","","","")
else:
    RED = '\033[31m'
    MAGENTA = '\033[35m'
    BLUE = '\033[34m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    LBLUE = '\033[36m'
    ENDC = '\033[0m'

COLUMNS = 80

def _get_terminal_width():
    '''
        Get the terminal width to adjust columns size
    '''
    global COLUMNS
    try:
        # (Slightly modified) http://stackoverflow.com/a/943921/91267
        p = os.popen('stty size', 'r')
        rows, columns = p.read().split()
        p.close()
        COLUMNS = int(columns)
    except:
        pass
_get_terminal_width()

class ETAQueue:

    def __init__(self, size, requests):
        self.size = size
        self.times = []
        self.pending_requests = requests

    def get_eta(self):
        mseconds_in_a_hour = 3600000
        mseconds_in_a_minute = 60000
        mseconds_in_a_second = 1000

        median = sum(self.times)//len(self.times)
        mseconds = self.pending_requests * median

        (r_hours, hours) = (mseconds % mseconds_in_a_hour, int(mseconds / mseconds_in_a_hour))
        (r_minutes, minutes) = (r_hours % mseconds_in_a_minute, int(r_hours / mseconds_in_a_minute))
        (r_seconds, seconds) = (r_minutes % mseconds_in_a_second, int(r_minutes / mseconds_in_a_second))

        if hours > 99:
            (hours, minutes, seconds) = ('NO','TE','ND')

        return "{0:2}h{1:2}m{2:2}s".format(hours, minutes, seconds)

    def set_time(self, time):
        self.pending_requests -= 1
        if len(self.times) == self.size:
            self.times.pop()
        self.times.append(time)

class Console:
    MIN_COLUMN_SIZE = 55
    eta_queue = None
    eta = "00h00m00s"
    show_full_path = False
    show_content_type = False

    @staticmethod
    def start_eta_queue(size, payload_length):
        Console.eta_queue = ETAQueue(size, payload_length)

    @staticmethod
    def header():
        header = os.linesep + "cod |    size    |  line  | time |" \
            + os.linesep + "----------------------------------" + os.linesep
        sys.stdout.write(header)

    @staticmethod
    def body(task):
        percentage = int(task.number * 100 / task.get_payload_length())
        target = task.get_complete_target()
        target = urlparse.urlsplit(target).path

        if task.location:
            target = target + " -> " + task.location
        if len(target) > COLUMNS - Console.MIN_COLUMN_SIZE:
            target = target[:abs(COLUMNS - Console.MIN_COLUMN_SIZE)]

        # If a task is valid means that is should be printed, so a proper
        # linesep will be printed
        linesep = ""
        if task.is_valid():
            linesep = os.linesep
        elif os.name == 'nt':
            return

        color = ""
        if task.response_code == "200":
            color = GREEN
        if task.response_code == "401" or task.response_code == "403":
            color = RED
        if task.response_code == "301" or task.response_code == "302":
            color = LBLUE
        if task.response_code.startswith('5') or task.response_code == '400':
            color = YELLOW
        if task.content_detected:
            color = MAGENTA

        to_format = color + "{1: ^3} | {2: >10} | {3: >6} | {4: >4} | {7} [{0: >2}%] - {5: ^9} - {6}" + ENDC
        to_format_without_progress = color + "{0: ^3} | {1: >10} | {2: >6} | {3: >4} | {5:^} {4}" + ENDC

        if sys.version_info[0] == 3:
            t_encode = target
        else:
            t_encode = target.encode('utf-8')

        # User wants to see full path
        if Console.show_full_path:
            t_encode = task.target[:-1] + t_encode

        # Fix three characters off by one on screen
        if percentage == 100:
            percentage = 99

        Console.eta_queue.set_time(task.response_time)
        if task.number % 10 == 0:
            Console.eta = Console.eta_queue.get_eta()

        # Show or not 'Content-Type'
        if Console.show_content_type:
            content_type = task.response_type + " |"
        else:
            content_type = ""

        # if an entry is about to be log, remove percentage and eta time
        if color:
            to_console = to_format_without_progress.format(task.response_code,
                                                    task.response_size,
                                                    task.number,
                                                    int(task.response_time),
                                                    t_encode, content_type)
        # print with progress
        else:
            to_console = to_format.format(percentage, task.response_code,
                                          task.response_size, task.number,
                                          int(task.response_time),
                                          Console.eta,
                                          t_encode, content_type)

        sys.stdout.write(to_console[:COLUMNS-2] + linesep)
        sys.stdout.flush()
        sys.stdout.write('\r')

        if not os.name == 'nt':
            sys.stdout.write("\x1b[0K")


utils/viewer.py
#!/usr/bin/env python
import sys
import sqlite3
import argparse

try:
    from asciitree import LeftAligned
except:
    print("install asciitree for -t option")

try:
    import urlparse
except:
    import urllib.parse as urlparse

RED = '\033[31m'
MAGENTA = '\033[35m'
BLUE = '\033[34m'
GREEN = '\033[32m'
YELLOW = '\033[33m'
LBLUE = '\033[36m'
ENDC = '\033[0m'

colors = {'200': GREEN,
          '302': BLUE,
          '403': RED,
          '500': YELLOW}


parser = argparse.ArgumentParser()
parser.add_argument('-p',
                    dest='project_name',
                    help='path to sqlite project database',
                    required=True)

parser.add_argument('-c',
                    dest='response_codes',
                    help='comma separate list of http codes responses to print',
                    required=False)

parser.add_argument('-u',
                    help='list the payload used on visited resources',
                    action='store_true')

parser.add_argument('-t',
                    dest='tree',
                    help='print a tree with the discoreved structure',
                    required=False,
                    action='store_true')

parser.add_argument('-e',
                    dest='banned_extensions',
                    help='comma separate list of extensions you do not want to see',
                    required=False)

parser.add_argument('-E',
                    dest='unbanned_extensions',
                    help='comma separate list of the extensions you only want to see (will include directories)',
                    required=False)

parser.add_argument('-s',
                    dest='size_filter',
                    help='comma separate list of size files you do not want to see',
                    required=False)


args = parser.parse_args()

project_name = args.project_name
response_codes = args.response_codes
banned_extensions = args.banned_extensions
unbanned_extensions = args.unbanned_extensions
size_filter = args.size_filter
used_payloads = args.u
tree = args.tree

if not response_codes:
    response_codes = ['200', '302', '403']

QUERY = "SELECT * FROM requests ORDER BY resource"
if used_payloads:
        QUERY = "SELECT url,payload FROM requests ORDER BY url"

try:
    with sqlite3.connect(project_name) as connection:
        cursor = connection.cursor()
        cursor.execute(QUERY)
        data = cursor.fetchall()

except Exception as e:
    print("Error opening database {}".format(project_name))
    print("Details: {}".format(e))
    sys.exit(-1)

def is_banned(resource):
    for banned in banned_extensions.split(','):
        if banned in resource.split('/')[-1]:
            return True
    return False

def is_ubanned(resource):
    if not '.' in resource.split('/')[-1]:
        return True
    for unbanned in unbanned_extensions.split(','):
        if unbanned in resource.split('/')[-1]:
            return True
    return False

def size_is_banned(size):
    if str(size) in size_filter.split(','):
        return True
    return False

def table():
    def comparator(a, b):
        if a.count('/') < b.count('/'):
            return -1
        if a.count('/') > b.count('/'):
            return 1
        if a.count('/') == b.count('/'):
            return 0

    if used_payloads:
        output = {}
        for k, v in data:
            if not output.has_key(k):
                output[k] = [v]
                continue
            output[k].append(v)
        for k in output.keys():
            print ("{}{}:".format(GREEN,k))
            output[k].sort()
            for v in set(output[k]):
                print ("        {}{}".format(BLUE,v))
        sys.exit()

    data.sort(cmp=comparator, key=lambda x: x[2] + x[3] + x[4])
    for line in data:
        (url, resource, extension, response_code, location, response_size) = \
            line[2], line[3], line[4], line[5], line[8], line[6]

        if banned_extensions and is_banned(resource):
            continue

        if unbanned_extensions and not is_ubanned(resource):
            continue

        if size_filter and size_is_banned(response_size):
            continue

        if response_code in response_codes:
            url = urlparse.urlparse(url)
            if not location:
                print("{}{} {:>8}: {}{}{}".format(colors[response_code],
                                                  response_code, response_size,
                                                  url.path, resource, extension))
            else:
                print("{}{} {:>8}: {}{}{} -> {}".format(colors[response_code],
                                                        response_code,
                                                        response_size, url.path,
                                                        resource, extension,
                                                        location))
def tree():
    if not LeftAligned:
        print("asciitree not installed!")
        sys.exit(-1)

    tr = LeftAligned()
    tree = {'/': {}}

    def _put_into_dict(root, resource):
        component = resource.split('/')[0]
        if not component:
            return
        if not component in root.keys():
            root[component] = {}
        resource = "/".join(resource.split('/')[1:])
        _put_into_dict(root[component], resource=resource)

    for entry in data:
        (resource, code, ext, response_size) = (entry[3], entry[5], entry[4], entry[6])

        if banned_extensions and is_banned(resource):
            continue

        if unbanned_extensions and not is_ubanned(resource):
            continue

        if size_filter and size_is_banned(response_size):
            continue
        if code in response_codes:
            if ext:
                resource = resource + ext
            _put_into_dict(tree['/'], resource)

    print(tr(tree))

if __name__ == '__main__':
    if args.tree:
        tree()
    else:
        table()


cansina.py
#!/usr/bin/env python
import sys
import os
import argparse
import time
import socket
import pickle
import tempfile

try:
    import urlparse
except:
    import urllib.parse as urlparse

try:
    import requests
    import urllib3
    urllib3.disable_warnings()
except ImportError:
    print("[CANSINA] Faltal Python module requests not found (install it with pip install --user requests)")
    sys.exit(1)

from datetime import timedelta

from core.visitor import Visitor
from core.payload import Payload
from core.dbmanager import DBManager
from core.printer import Console
from core.resumer import Resumer
from plugins.robots import process_robots
from plugins.inspector import Inspector

# Workaround for Python SSL Insecure*Warnings
ver = sys.version_info
if ver.major == 2 and ver.minor == 7 and ver.micro < 9:
    print("[!] Python < 2.7.9 - SSL Warnings disabled")
    from requests.packages.urllib3.exceptions import InsecureRequestWarning
    from requests.packages.urllib3.exceptions import InsecurePlatformWarning
    from requests.packages.urllib3.exceptions import SNIMissingWarning
    requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
    requests.packages.urllib3.disable_warnings(InsecurePlatformWarning)
    requests.packages.urllib3.disable_warnings(SNIMissingWarning)
del ver

try:
    raw_input          # Python 2
except NameError:
    raw_input = input  # Python 3

#   Default options
#
USER_AGENT = "Mozilla/5.0 (Windows; U; MSIE 10.0; Windows NT 9.0; en-EN)"
THREADS = 4

#
#   Utility functions
#


def _check_domain(target_url):
    """Get the target url from the user, clean and return it"""

    domain = urlparse.urlparse(target_url).hostname
    print("Resolving " + domain)
    try:
        if socket.gethostbyname(domain):
            pass
    except Exception:
        print("Domain doesn't resolve. Check URL")
        sys.exit(1)


def _prepare_target(target_url):
    """Examine target url compliance adding default handle (http://) and look for a final /"""

    if target_url.startswith('http://') or target_url.startswith('https://'):
        pass
    else:
        target_url = 'http://' + target_url
    if target_url.endswith('/') or '***' in target_url:
        pass
    else:
        target_url += '/'
    _check_domain(target_url)
    return target_url


def _prepare_proxies(proxies):
    """It takes a list of proxies and returns a dictionary"""

    if proxies:
        proxies_dict = {}
        for proxy_item in proxies:
            if proxy_item.startswith('http://'):
                proxies_dict['http'] = proxy_item
            elif proxy_item.startswith('https://'):
                proxies_dict['https'] = proxy_item
        return proxies_dict
    return {}


def _make_cookie_jar(_cookies):
    d = dict()
    c = _cookies.split(',')
    if c[0] == '':
        return d
    for entry in c:
        key, value = entry.split(':')
        d[key] = value
    return d

#
# Parsing program options
#
usage = "cansina.py -u url -p payload [options]"
description = '''
Cansina is a web content discovery tool.
It makes requests and analyze the responses trying to figure out whether the
resource is or not accessible.
'''
epilog = "License, requests, etc: https://github.com/deibit/cansina"

parser = argparse.ArgumentParser(
    usage=usage, description=description, epilog=epilog)
parser.add_argument('-A', dest='authentication',
        help="Basic Authentication (e.g: user:password)", default=False)
parser.add_argument('-C', dest='cookies', help="your cookies (e.g: key:value)", default="")
parser.add_argument('-D', dest='autodiscriminator',
                    help="Check for fake 404 (warning: machine decision)", action="store_true", default=False)
parser.add_argument('-H', dest='request_type',
                    help="Make HTTP HEAD requests", action="store_true")
parser.add_argument('-P', dest='proxies',
                    help="Set a http and/or https proxy (ex: http://127.0.0.1:8080,https://...", default="")
parser.add_argument('-S', dest='remove_slash',
                    help="Remove ending slash for payloads", default=False, action="store_true")
parser.add_argument('-T', dest='request_delay',
        help="Time (a float number, e.g: 0.25 or 1.75) between requests", default=0)
parser.add_argument('-U', dest='uppercase',
                    help="Make payload requests upper-case", action="store_true", default=False)
parser.add_argument('-a', dest='user_agent',
                    help="The preferred user-agent (default provided)", default=USER_AGENT)
parser.add_argument('-b', dest='banned',
                    help="List of banned response codes", default="404")
parser.add_argument('-B', dest='unbanned',
                    help="List of unbanned response codes, mark all response as invalid without unbanned response codes, higher priority than banned", default="")
parser.add_argument('-c', dest='content',
                    help="Inspect content looking for a particular string", default="")
parser.add_argument('-d', dest='discriminator',
                    help="If this string if found it will be treated as a 404", default=None)
parser.add_argument('-e', dest='extension',
                    help="Extension list to use e.g: php,asp,...(default none)", default="")
parser.add_argument('-p', dest='payload',
                    help="A single file, a file with filenames (.payload) or a directory (will do *.txt)", default=None)
parser.add_argument('-s', dest='size_discriminator',
                    help="Will skip pages with this size in bytes (or a list of sizes 0,500,1500...)", default=False)
parser.add_argument('-t', dest='threads', type=int,
                    help="Number of threads (default 4)", default=THREADS)
parser.add_argument('-u', dest='target',
                    help="Target url", default=None)
parser.add_argument('-r', dest='resume',
                    help="Resume a session", default=False)
parser.add_argument('-R', dest="parse_robots", action="store_true",
                    help="Parse robots.txt and check its contents", default=False)
parser.add_argument('--recursive', dest="recursive",
                    help="Recursive descend on path directories", default=False, action="store_true")
parser.add_argument('--persist', dest="persist",
                    help="Use HTTP persistent connections", default=False, action="store_true")
parser.add_argument('--full-path', dest="full_path",
                    help="Show full path instead of only resources", default=False, action="store_true")
parser.add_argument('--show-type', dest="show_content_type",
                            help="Show content-type in results", default=False, action="store_true")
parser.add_argument('--no-follow', dest="allow_redirects",
                            help="Do not follow redirections", default=True, action="store_false")

args = parser.parse_args()

# Initialize a Resumer object
resumer = Resumer(args, 0)
resume = args.resume
# If we are ressuming a former session revive last args object
if resume:
    try:
        with open(resume) as f:
            resumer = pickle.load(f)
            args = resumer.get_args()
    except:
        sys.stdout.write("[!] Could not load a correct resume file, sorry.")
        sys.exit()

# Target check and preparation
if not args.target:
    print("[!] You need to specify a target")
    parser.print_help()
    sys.exit()
target = _prepare_target(args.target)

recursive = args.recursive
if args.recursive:
    print("Recursive requests on")

# Persistent connections
persist = args.persist
if persist:
    print("Activated persistent connections")

# Misc options
extension = args.extension.split(',')
threads = int(args.threads)
banned_response_codes = args.banned.split(',')
unbanned_response_codes = args.unbanned.split(',')
user_agent = args.user_agent
proxy = _prepare_proxies(args.proxies.split(','))
cookies = args.cookies

# HEAD / GET requests
request_type = args.request_type
if request_type:
    print("HTTP HEAD requests")
    request_type = "HEAD"
else:
    print("HTTP GET requests")
    request_type = "GET"

# Content inspecting
content = args.content
if content:
    print("Content inspection selected")
    if request_type == "HEAD":
        print("[!] head request render Content Inspection useless")

# Remove slash (probably an ancient option)
remove_slash = args.remove_slash
if remove_slash:
    print("Requests without last /")

# Discriminator option
discriminator = args.discriminator
if discriminator:
    print("Discriminator active")
    if request_type == "HEAD":
        print("[!] head requests renders discriminator by content useless")

# Autodiscriminator (probably deprecated by future diagnostic subsystem)
autodiscriminator = args.autodiscriminator
autodiscriminator_location = None
autodiscriminator_md5 = None
if autodiscriminator:
    print("Launching autodiscriminator")
    i = Inspector(target)
    (result, notfound_type) = i.check_this()
    if notfound_type == Inspector.TEST404_URL:
        autodiscriminator_location = result
        print("404 ---> 302 ----> " + autodiscriminator_location)
    elif notfound_type == Inspector.TEST404_MD5:
        autodiscriminator_md5 = result
        print("404 ---> PAGE_MD5 ----> " + autodiscriminator_md5)

# Misc user information
print("Banned response codes: %s" % " ".join(banned_response_codes))
if not unbanned_response_codes[0] == '':
    print("unBanned response codes: %s" % " ".join(unbanned_response_codes))
if not extension == ['']:
    print("Extensions to probe: %s" % " ".join(extension))

# Uppercase
uppercase = args.uppercase
if uppercase:
    print("All resource requests will be done in uppercase")

# Misc options
request_delay = args.request_delay
authentication = args.authentication
size_discriminator = args.size_discriminator

# Payload options
# FIXME: This design is garbage
payload = None
parse_robots = args.parse_robots
remove_slash = True
if parse_robots:
    robots_content = process_robots(target)
    if not robots_content:
        print("[!] robots.txt not found")
        sys.exit()
    print("Reaped %s entries" % (len(robots_content)))
    print("Using robots.txt as payload")
    payload_filename = robots_content
else:
    payload_filename = args.payload
    if not payload_filename:
        print("[!] You have to specify a payload")
        parser.print_help()
        sys.exit()
    print("Using payload: %s" % payload_filename)
    print("Generating payloads...")

payload = Payload(target, payload_filename, resumer)
print("Spawning %s threads " % threads)

#
# Payload queue configuration
#
payload.set_extensions(extension)
payload.set_remove_slash(remove_slash)
payload.set_uppercase(uppercase)
payload.set_banned_response_codes(banned_response_codes)
payload.set_unbanned_response_codes(unbanned_response_codes)
payload.set_content(content)
payload.set_recursive(recursive)

#
# Manager queue configuration
#
database_name = urlparse.urlparse(target).hostname
manager = DBManager(database_name)

#
# Configure Visitor Objects
#
Visitor.set_authentication(authentication)
Visitor.set_banned_location(autodiscriminator_location)
Visitor.set_banned_md5(autodiscriminator_md5)
Visitor.set_delay(request_delay)
Visitor.set_discriminator(discriminator)
Visitor.set_proxy(proxy)
Visitor.set_requests(request_type)
Visitor.set_size_discriminator(size_discriminator)
Visitor.set_user_agent(user_agent)
Visitor.set_persist(persist)
Visitor.allow_redirects(args.allow_redirects)

try:
    cookie_jar = _make_cookie_jar(cookies)
    Visitor.set_cookies(cookie_jar)
    if cookie_jar:
        print("Using cookies")
except:
    print("Error setting cookies. Review cookie string (key:value,key:value...)")
    sys.exit()

payload_queue = payload.get_queue()
total_requests = payload.get_total_requests()
print("Total requests %s  (aprox: %s / thread)" %
      (total_requests, int(total_requests / threads)))
#
# Create the thread_pool and start the daemonized threads
#
thread_pool = []
for visitor_id in range(0, threads):
    v = Visitor(visitor_id, payload_queue, manager)
    thread_pool.append(v)
    v.daemon = True
    v.start()

# Select if full path is prefered
full_path = args.full_path

# Select if user wants content-type print
show_content_type = args.show_content_type

#
#   Run the main thread until manager exhaust all tasks
#
time_before_running = time.time()
Console.start_eta_queue(30, total_requests)
Console.show_full_path = full_path
Console.show_content_type = show_content_type
Console.header()

# Main loop
try:
    while True:
        visitors_alive = any([visitor.is_alive() for visitor in thread_pool])
        if not manager.get_a_task(visitors_alive):
            break
except KeyboardInterrupt:
    sys.stdout.write(os.linesep + "Waiting for threads to stop...")
    Visitor.kill()
    resp = raw_input(os.linesep + "Keep a resume file? [y/N] ")
    if resp == 'y':
        resumer.set_line(payload_queue.get().get_number())
        with open("resume_file_" + time.strftime("%d_%m_%y_%H_%M", time.localtime()), 'w') as f:
            pickle.dump(resumer, f)
except Exception as e:
    import traceback as tb
    sys.stderr.write("Unknown exception: %s" % e)
    print(tb.print_tb(sys.exc_info()[2]))

sys.stdout.write("Finishing...")

time_after_running = time.time()
delta = round(timedelta(seconds=(time_after_running -
                                 time_before_running)).total_seconds(), 2)
print("Task took %i seconds" % delta)

sys.stdout.flush()
sys.exit()
